sycl/doc/extensions/experimental/sycl_ext_oneapi_compat.asciidoc

```diff
@@ -0,0 +1,1346 @@
+  int id_x = compat::local_id::x();
+  int id_y = compat::local_id::y();
+  As[id_y][id_x] = id_x * BLOCK_SIZE + id_y;
+  compat::wg_barrier();
+  int val = As[BLOCK_SIZE - id_y - 1][BLOCK_SIZE - id_x - 1];
+  d_A[compat::global_id::y() * BLOCK_SIZE + compat::global_id::x()] = val;
+}
+```
+
+==== launch<function>
+
+_sycl_ext_oneapi_compat_ provides a kernel `launch` interface which accepts a
+function that executes on the device (a.k.a "kernel") instead of a
+lambda/functor. It can be called either by using a pair of "teams"/"blocks" and
+"threads", from OpenMP/CUDA terminology, or using a `sycl::nd_range`. The kernel
+function is passed via the template argument `F`.
+
+Various overloads for `launch<function>` exist to permit the user to launch on a
+specific `queue`, or to define dynamically sized local memory.
+
+``` c++
+namespace sycl::ext::oneapi::experimental::compat {
+
+template <auto F, typename... Args>
+sycl::event launch(const dim3 &grid, const dim3 &threads, Args... args);
+
+template <auto F, int Dim, typename... Args>
+sycl::event launch(const sycl::nd_range<Dim> &range, Args... args);
+
+template <auto F, int Dim, typename... Args>
+sycl::event launch(const sycl::nd_range<Dim> &range,
+                   sycl::queue q, Args... args);
+
+template <auto F, typename... Args>
+sycl::event launch(const dim3 &grid, const dim3 &threads,
+                   sycl::queue q, Args... args);
+
+template <auto F, int Dim, typename... Args>
+sycl::event launch(const sycl::nd_range<Dim> &range, size_t num_local_elements,
+                   sycl::queue q, Args... args);
+
+template <auto F, int Dim, typename... Args>
+sycl::event launch(const sycl::nd_range<Dim> &range, size_t num_local_elements,
+                   Args... args);
+
+template <auto F, typename... Args>
+sycl::event launch(const dim3 &grid, const dim3 &threads,
+                   size_t num_local_elements, sycl::queue q, Args... args);
+
+template <auto F, typename... Args>
+sycl::event launch(const dim3 &grid, const dim3 &threads,
+                   size_t num_local_elements, Args... args);
+
+} // sycl::ext::oneapi::experimental::compat
+```
+
+For example, if the user had an existing function named `vectorAdd` to execute
+on a device such as follows:
+
+``` c++
+void vectorAdd(const float *A, const float *B, float *C, int numElements);
+```
+
+using this compatibility extension the user can call it as follows:
+
+``` c++
+compat::launch<vectorAdd>(blocksPerGrid, threadsPerBlock, d_A, d_B, d_C,
+                          numElements);
+```
+
+which would be equivalent to the following call using a `sycl::nd_range`:
+
+``` c++
+auto range = sycl::nd_range<3>{blocksPerGrid * threadsPerBlock,
+                               threadsPerBlock};
+compat::launch<vectorAdd>(range, d_A, d_B, d_C, numElements);
+```
+
+For dynamic local memory allocation, `launch<function>` deduces the typing of
+the device's local memory from the kernel function signature.
```

Comment:
> Why does the kernel's function signature need to be modified? The example void local_mem_2d(int *d_A) function doesn't include a parameter for local memory.


There are two distinct interfaces to allocate local memory  using `SYCLcompat`. The one you are mentioning `local_mem_2d` uses the current implementation of _sycl_ext_oneapi_local_memory_ to provide statically sized memory allocation. In this case, the kernel function signature is not modified because it does not need an additional argument to access the local memory, the memory is defined inside the function.

The second interface to allocate local memory is used to provide dynamically sized memory through the `launch<F>` function. The user can define the size at runtime, and internally, `SYCLcompat` will inject a pointer to the local memory location as an additional argument of the kernel function. That is the reason to modify the kernel function signature.

We have modified the proposal to improve readability, since other users may arrive to the same conclusion. Sorry about the misunderstanding.